### DISCUSSION TITLE
Chord Suggestion Tooltip

### DISCUSSION THREAD

**Discussion by hiddengarden** [psoukie/zipchord#93] [Ideas]
How about leveraging text prediction in windows or otherwise to display the relevant chord suggestion as a learning feature?
Ideally, it should display next to the whole words.

**Comment by psoukie** [comment]
Thank you for the suggestion! I really like the idea of letting the user set the display of hints to no hints, shown in tooltips (next to cursor), and shown in OSD.

The "next to cursor" part seems tricky to detect in a lot of applications (like Microsoft Word) but there seem to be workarounds for that.

In terms of what the hint should be hinting at: my focus for ZipChord has been on it being a tool for touch typing (i.e. no need to look at the screen) and on what you want to express (rather than suggestions). So the hints are definitely meant as a learning tool, and the idea was to show the user a shortcut for the word that was just typed manually.

To make it work as autocomplete hints, ZipChord would need to know (and send to another service) more than just the last typed word, and show the text of suggestions together with its shortcuts, which might clash with autocomplete built into the editor being used.

On a related note, to avoid the hints being annoying or distracting, I think they should also not be displaying many hints in a row. If I start typing normally (without chords/shorthands), they should temporarily disappear. (So let's say a timer of a few seconds.)

I'd love to hear more thoughts on this.

**Reply by Nooby-001** [reply]
I like the hints the way they are right now, version 2.0, I think. My thing is to use mnemonics for autocomplete (murc is America, murca is United States, murcas is the whole USA, etc) so tips are really useful to memorize.
Don't see the point of "next to cursor" since you're looking at the screen anyway so i'm ok with the tips displaying around the center of the monitor.
Ideas: Maybe a way to change the tip's size/color, draw them inside a box? 
Sound, A soft beep when there's a chord/autocomplete available.

**Reply by hiddengarden** [reply]
version 2.0 already! I use 1.7.1 :)

**Reply by hiddengarden** [reply]
Just installed the 2.0 beta. I like those hints too!

**Comment by psoukie** [comment]
> How about leveraging text prediction in windows or otherwise to display the relevant chord suggestion as a learning feature? Ideally, it should display next to the whole words.

This is now implemented in the [source script](https://github.com/psoukie/zipchord/blob/main/source/zipchord.ahk) (not released yet). I tried to make compromises so it would be positioned decently across different apps (browser, Word, Notepad, VS Code).

> Maybe a way to change the tip's size/color, draw them inside a box?

I don't want to overwhelm the settings with too many things. What kind of look / style would you recommend? (I have an issue for improving the look of the OSD under #89)

 > On a related note, to avoid the hints being annoying or distracting, I think they should also not be displaying many hints in a row. 

I put some ideas on this into #97 -- have three options: show all hints; "normal" (pause if a 3rd tip would be shown within a 30-second period, and resume after 2 minutes), relaxed (pause if a 2nd tip would be shown within a 30-second period, and resume after 5 minutes). 

I'm open to suggestions and feedback on all these topics.

**Comment by hiddengarden** [comment]
Thank you for seeing value in my suggestion! The feature I aimed to describe, somewhat clumsy, is similar to word prediction only by its appearance. It is essential to have only one app doing the same on my  PC.  One of the features is turned off immediately.

Prediction and hints availability for the next word, as you describe it, to me can be redundant because it could quickly become overwhelming to the user. Also, when I write, dealing with prediction is sometimes disruptive. 

I prefer a small tip showing the chord relevant to the word I just typed, so I can reinforce memorising the chord in the dictionary. A small tooltip with the chord that pops at the top of the cursor 2-5 seconds after the space (period, comma, colon, etc.) has indicated the completion of the word. Then a 7-10 seconds fadeout would be nice. 

![image](https://user-images.githubusercontent.com/115948281/211187463-6b258553-63af-4355-a9de-41f14a17cc8b.png)

This shouldn't apply to words entered using a chord.
In case of chord entry, some quick encouragement may be fun to show, like:

- COOL!
- WOW!
- YAY!
- Good job!
- Well done!

One of those at random will be best.

Also, I just noticed I'm using 1.7.1 and have yet to see 2.0. I'll make sure to update before posting more suggestions.

Many Thanks!

**Reply by hiddengarden** [reply]
To implement Gamification into the product:

You can take this one step further by starting to count chord usage and introducing hit count similar to  your favourite fighting game:

- Combo!
- X3 Combo!
- X4 Combo!
- Pentakill
- Star
- Domination!
- Superiority!
- Annihilation!

Subject to a discussion, but you get the idea. You can even consider a scoreboard within the app settings menu, which should display the achievement count per day and then per week. Leaderboards should be out of scope unless you want users to share those statistics and everything that goes with it. 

Please let me know if you need more input. I am open to answering more specific questions and suggesting designs around that, although Timing is not my virtue.

Cheers!

**Reply by psoukie** [reply]
Great minds think alike :) The suggestions were part of 2.0, and I implemented them as tooltips on Friday. The features are now available in the [Release Candidate](https://github.com/psoukie/zipchord/releases/tag/2.0.0-rc.1). 

I'll track the gamification as a possible addition.

**Reply by psoukie** [reply]
> In case of chord entry, some quick encouragement may be fun to show...

As part of the learning / gamification. Since  I learn by doing, I find that when I get a tooltip for a shortcut, I am more successful when I try it immediately (to start building muscle memory). So it would also be nice if in that mode, when you type a word normally and get a hint, there was a "Try it now" feature so that if you then press the shortcut for that word, it wouldn't double your output, but just showed a confirmation / positive message.

**Reply by hiddengarden** [reply]
This will complete the experience. Looking forward to seeing it in action!
Would you be inclined to run a small before-and-after study on the gamification feature? I'll be very curious to look at the engagement and download numbers and publish them in my LinkedIn profile.

Regarding Confirmation / Positive message: How about a sound of a cheering crowd? :)
This is another opportunity to count and come up with a cognition score where a point is awarded for success, and one is taken away for failure. Then 10 points for consecutively chording the same word successfully five times a row. Then stop monitoring the same word. 

I realise this may involve building a database for all chords, but in the long run, this can help solidify the predefined chords and boil them down to create a consensus optimising efficiency. Turn this into a skill instead of a group of people with different custom chord definitions.
Again hats off to you, and much respect for your project!

Let me know about the study.

**Reply by psoukie** [reply]
Thank you. Since this tool assists with typing, I want to be careful about features that send any data over the internet. But I think there will be a way to have something where people could agree to send statistics, or -- to make it completely safe -- just to generate a file they could email or post somewhere.

Because I'm not tracking anything online, I honestly do not have an idea how many people might be using this tool. Hopefully, with the version 2.0 coming out, it will pick up. There's one other feature I want to implement first, and it's been busy with my day job, but I'll definitely get to the features supporting learning the chording soon. I created an issue #113 to track this.

For standardized as opposed to custom -- I find it easy when I define my own chords because there's some logic to it. (I'm also one of the few crazy people typing on Dvorak layout, so some of my QWERTY default chord suggestions might  be subpar.) But yes, developing a more complete dictionary that people can use -- or adapt -- or throw away and start their own ... would be nice.

I'm definitely open to other people contributing to this project.

**Reply by psoukie** [reply]
After a long pause, I was able to resume, and now started implementing some of these ideas. Thank you for suggesting them!

**Comment by hiddengarden** [comment]
I feel you! Colemak user over here.
:) I didn't make the connection between the layout and the shorthand
awkwardness, but now it explains everything. Lol

I too believe that customising should be a permanent backbone feature. In
fact, replacing that would be a deal breaker for me.
If it wasn't there, I probably wouldn't take it on to learn at all.

I completely understand you must follow your plan and it makes sense! I
would absolutely love to help with documentation as a labour of love, but
there is too much on my plate right now. Maybe I'll be able to find more
time next month and really contribute with something more than ideas.

Cheers!
